blog/_src/posts/2016-07-25-tutorial-zero-to-sixty-in-racket.scrbl

```diff
@@ -0,0 +1,698 @@
+      (append (rest words) (list (first words))))
+
+    (module+ test
+      (check-equal?
+        (circular-shift '("A" "B" "C"))
+        '("B" "C" "A")))
+
+    (define (alphabetize all-shifts)
+      (sort all-shifts shift<?))
+
+    (module+ test
+      (check-equal?
+        (alphabetize '(("racket" "is")
+                       ("as")
+                       ("racket" "does")))
+        '(("as")
+          ("racket" "does")
+          ("racket" "is"))))
+))
+
+
+@section{Recognizing Patterns, Avoiding Repetition}
+
+Every unit test needs to go through @hyperlink["http://docs.racket-lang.org/rackunit/api.html#%28def._%28%28lib._rackunit%2Fmain..rkt%29._check-equal~3f%29%29"]{check-equal?}.
+
+@(begin
+  #reader scribble/comment-reader
+  (racketblock
+
+    (module+ test
+      (check-equal?
+        (kwik-split '())
+        '())
+      (check-equal?
+        (kwik-split '("hello    world"))
+        '(("hello" "world")))
+      (check-equal?
+        (kwik-split '(" lost " " in " "space"))
+        '(("lost") ("in") ("space")))
+      (check-equal?
+        (kwik-split '("something"))
+        '()))
+))
+
+These tests follow a simple pattern that we can express as a @emph{syntax rule}.
+
+@(begin
+  #reader scribble/comment-reader
+  (racketblock
+
+    (module+ test
+      (define-syntax-rule (check-equal?* [i o] ...)
+        (begin
+          (check-equal? i o)
+          ...))
+
+      (check-equal?*
+        [(kwik-split '())
+         '()]
+        [(kwik-split '("hello    world"))
+         '(("hello" "world"))]
+        [(kwik-split '(" out " " in " "the ozone"))
+         '(("out") ("in") ("the" "ozone"))]
+        [(kwik-split '("something"))
+         '()]))
+))
+
+The @racket[...] are not pseudocode!
+They denote Kleene-star repetition, like a sextile (@racket{*}) in a regular expression.
+In this case, the input pattern is just a list of two S-expressions, @racket[i] and
+ @racket[o].
+Uses of @racket[i] and @racket[o] must be followed by one @racket[...] to splice
+ the captured S-expressions into the result.
+
+Many languages offer higher-order functions and polymorphism to abstract common
+ behaviors.
+Syntax extensions are a different way to avoid repeating yourself.
+After 30 years, we are still discovering what syntax extensions are useful for.
+
+See this @hyperlink["https://groups.google.com/forum/#!topic/racket-users/ss20lwfUhjs/discussion"]{recent Racket mailing list post} for some applications.
+
+
+@section{Adding Static Types}
+
+Changing the @hyperlink["http://docs.racket-lang.org/reference/reader.html#%28idx._%28gentag._79._%28lib._scribblings%2Freference%2Freference..scrbl%29%29%29"]{@tt{#lang}} line to @racket[typed/racket] adds static type-checking to our program.
+If we only change the language and run the code as-is, there will be type errors.
+But we can use submodules again to incrementally arm our design with types.
+
+@codeblock{
+   #lang racket
+
+    (module t typed/racket
```

Comment:
you can also use typed regions for this https://plt.eecs.northwestern.edu/snapshots/current/doc/ts-reference/Typed_Regions.html?q=typed-region